### PR TITLE
mem: Improve protection

### PR DESCRIPTION
**Pull request by Macdu**
Fix many issues concerning our memory protection, mainly:
- Out of bound iterator reads and writes
- remove all the const_casts
- some speed optimizations
- better debugging and logging of windows / unix errors
- fix memory unprotecting happening when it shouldn't
- replace all size_t by uint32_t: this is confusing and we are emulating a 32-bit address range anyway

### CODE REVIEW

File: vita3k/mem/src/mem.cpp [hunk 1]
```diff
@@ -202,23 +215,13 @@ Address alloc(MemState &state, size_t size, const char *name, unsigned int align
     return align_addr;
 }
 
-static void align_to_page(MemState &state, ProtectSegmentInfo &protect) {
-    const Address end = align(protect.addr + protect.size, state.page_size);
-    const Address addr = align_down(protect.addr, state.page_size);
-    const size_t size = end - addr;
-    protect.addr = addr;
-    protect.size = size;
+static void align_to_page(MemState &state, Address &addr, Address &size) {
```
Comment (bookmist) [inline vita3k/mem/src/mem.cpp:218]: It's a bit confusing that addr and size have same type, it looks like size is also a guest pointer. (My suggestion is Address &addr, uint32_t &size)

File: vita3k/mem/src/mem.cpp [hunk 2]
```diff
@@ -227,54 +230,47 @@ void unprotect_inner(MemState &state, Address addr, size_t size) {
...
 #endif
 }
 
-void protect_inner(MemState &state, Address addr, size_t size, const std::uint32_t perm) {
+void protect_inner(MemState &state, Address addr, uint32_t size, const uint32_t perm) {
     uint8_t *addr_ptr = state.use_page_table ? state.page_table[addr / KiB(4)] : state.memory.get();
 
 #ifdef WIN32
     DWORD old_protect = 0;
     const BOOL ret = VirtualProtect(&addr_ptr[addr], size - 1, (perm == MEM_PERM_NONE) ? PAGE_NOACCESS : ((perm == MEM_PERM_READONLY) ? PAGE_READONLY : PAGE_READWRITE), &old_protect);
```
Comment (bookmist) [inline vita3k/mem/src/mem.cpp:245]: It could be good to make MEM_PERM_* enum named and use its type instead of uint32_t

**Review by bookmist — DISMISSED** [review]
Great. I hope it fix a lot of uint32/uint64 conversion warnings.

File: vita3k/mem/src/mem.cpp (hunk 1 above)
Comment (Macdu) [reply vita3k/mem/src/mem.cpp:218]: Indeed, you are right

File: vita3k/mem/src/mem.cpp (hunk 2 above)
Comment (Macdu) [reply vita3k/mem/src/mem.cpp]: done

**Review by bookmist — APPROVED** [review]

**Pull request merged**